### ISSUE TITLE
TypeVarTuple equality check for tuples

### ISSUE THREAD

**Issue by cdce8p** [microsoft/pyright#6888]
**Describe the bug**
https://peps.python.org/pep-0646/#type-variable-tuple-equality

**Code or Screenshots**
Code example from the PEP, see link above.

```python
from typing import TypeVarTuple
Ts = TypeVarTuple("Ts")

def foo(arg1: tuple[*Ts], arg2: tuple[*Ts]): ...

a = (0,)
b = ('0',)
foo(a, b)  # This should be an error
```

**VS Code extension or command-line**
pyright 1.1.344

**Comment by erictraut** [comment]
Pyright is working correctly here.

The above code should type check without error for the same reason that the following code type checks without error:

```python
T = TypeVar("T")

def bar(arg1: T, arg2: T) -> T:
    ...

reveal_type(bar(0, "0")) # int | str
```

The constraint solver is able to find a solution for `Ts` that meets all of the requirements in your code.

```python
Ts = TypeVarTuple("Ts")

def foo(arg1: tuple[*Ts], arg2: tuple[*Ts]) -> tuple[*Ts]:
    ...

a = (0,)
b = ("0",)
reveal_type(foo(a, b))  # tuple[int | str]
```

**Comment by cdce8p** [comment]
That's true, however the PEP explicitly mentioned it is not allowed. That's why I opened the issue.
> We do not allow this; type unions may not appear within the Tuple. If a type variable tuple appears in multiple places in a signature, the types must match exactly (the list of type parameters must be the same length, and the type parameters themselves must be identical)

https://peps.python.org/pep-0646/#type-variable-tuple-equality

**Comment by erictraut** [comment]
Ah interesting. That part of the PEP was apparently added late in the PEP's evolution, long after I implemented support for it in pyright. If I knew the authors had added that provision, I would have pushed back because it's inconsistent with other types of generics in the Python type system. 

OK, I'll re-open the issue.

**Comment by erictraut** [comment]
Thinking about this more, I don't think this rule makes any sense. I'm going to push to amend the typing spec to fix this "spec bug".

With the current definition, the following is indeterminate:
Ts = TypeVarTuple("Ts")

def foo(arg1: tuple[*Ts], arg2: tuple[*Ts]) -> tuple[*Ts]: ...
foo((0, ), (0.0, ))
```

Should this produce `tuple[float]` or generate an error as the current spec would seem to imply?

How about `foo((0.0, ), (0, ))`? Is this now OK because the leftmost tuple is assumed to be captured first by `Ts`, and the second tuple is compatible with it? This would be the first and only place in the typing spec where order of type evaluation determines correctness, which is very problematic.

How about `foo((0, ), (1, ))`? Should this generate an error because the first tuple is `tuple[Literal[0]]` and the second tuple is incompatible with that type?

I'm going to leave this issue open for now, but I don't plan to "fix" it in pyright because I think the spec is buggy in this case.

**Comment by hmc-cs-mdrissi** [comment]
The underlying motivation for this rule is for cases like,

```python
Ts = TypeVarTuple("Ts")

class Tensor(Generic[Ts]):
  ...
def add_tensors(x: Tensor[*Ts], y: Tensor[*Ts]):
  ...

a = Tensor[Literal[1], Literal[2]](...)
b = Tensor[Literal[1], Literal[2]](...)
c = Tensor[Literal[1], Literal[3]](...)

add_tensors(a, b) # Good
add_tensors(a, c) # Bad as shape mismatch. Undesired to have Union like Literal[2, 3] appear as context for math/matrices wants exact match.
```

The complexity is with ambiguous/inferred types especially given type inference allows multiple answers for a variable depending on type checker today.

**Comment by erictraut** [comment]
I understand the motivation, but whoever wrote that part of the spec didn't understand some fundamental aspects of type theory, constraint solving, and type inference.

There may be other ways we can address this use case, but the current spec is fundamentally flawed IMO and must be changed.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.348, which was just published. It will also be included in a future release of pylance.
